### PR TITLE
return partial matches when default latest tag is unavailable

### PR DESCRIPTION
**Pull request by bparees**
fixes bug https://bugzilla.redhat.com/show_bug.cgi?id=1419849

### CODE REVIEW

**Comment by bparees** [comment]
@jim-minter ptal.  the issue here was that the tag followed by the latest tag had failed to import.

This meant that when the user specified no specific tag, and we defaulted the tag to "latest" and then attempted to resolve the follow, we'd ultimately conclude that we couldn't resolve it to anything useful, so we'd go down the "find tags that do exist and return those as partial matches" code path, except that in this case the resulting score was "0"(perfect match).

As a result we'd add the match and then use it, w/o ever having looked up the imagemetadata, which meant we didn't create a service since we never saw the exposed ports metadata associated w/ the image.

With this change, we should be treating all matches as partial matches unless either:

1) it's exactly the tag the user specified
2) it's a match we derived from the latest tag (and the user didn't specify a tag, or specified latest)

**Comment by bparees** [comment]
[test]

File: pkg/generate/app/imagestreamlookup.go [hunk 1]
```diff
@@ -162,10 +162,14 @@ func (r ImageStreamSearcher) Search(precise bool, terms ...string) (ComponentMat
...
-						tagScore := score
+						// at best this is a partial match situation.  The user didn't
+						// specify a tag, so we tried "latest" but could not find an image associated
+						// with the latest tag (or one that is followed by the latest tag), or
+						// they specified a tag that we could not find.
+						tagScore := score + 0.1
 						if !followTag {
+							// If the user specified a tag in their search string (followTag == false),
+							// then score this match lower.
 							tagScore += 0.5
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: @jim-minter i have half a mind to just drop this if statement and always set tagScore:=score+0.5, i don't think there's a meaningful distinction between score=0.1 and score=0.6, nor am i convinced there's a good justification for why one should be 0.1 and the other should be 0.6, but let me know what you think.

**Comment by bparees** [comment]
[test]

Ben Parees | OpenShift

File: pkg/generate/app/imagestreamlookup.go (hunk 1 above)
Comment (jim-minter) [inline pkg/generate/app/imagestreamlookup.go]: @bparees I agree with always setting tagScore := score + 0.5 regardless of the value of followTag.  I think the intended behaviour it changes is when someone does `oc new-app builder~repo`, where no tag is specified as part of `builder` and no "latest" image is known to us.  In that case, currently we'll go with any other tag we find.  I think your proposal would mean we'd print the options and stop, which doesn't seem like a bad plan.

**Comment by jim-minter** [comment]
@bparees I haven't understood why it was that we didn't look up the imagedata (and therefore didn't create a service) in this case?  Is there a quick answer?

**Comment by jim-minter** [comment]
@bparees finally, I wonder whether this is worthy of a test case.

File: pkg/generate/app/imagestreamlookup.go (hunk 1 above)
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: >  In that case, currently we'll go with any other tag we find.

Right, that was the previous behavior if only one tag was found.  if more than one was found, we used to print them all and ask for more details (and we still will). 

> I think your proposal would mean we'd print the options and stop, which doesn't seem like a bad plan.

yeah, that's already what this code change does and i think it's the right behavior.

**Comment by bparees** [comment]
>  I haven't understood why it was that we didn't look up the imagedata (and therefore didn't create a service) in this case? Is there a quick answer?

There are two ways matches are added.  For "exact" matches (score 0), we lookup the metadata and add it:
https://github.com/openshift/origin/blob/master/pkg/generate/app/imagestreamlookup.go#L179-L190

for "inexact" matches (which is the path we were going through here where we didn't find a latest tag, but we found another tag and incorrectly scored it as an exact match), we do not bother looking up metadata because the assumption is that for inexact matches we're just going to print it out to the user not do actual generation:
https://github.com/openshift/origin/blob/master/pkg/generate/app/imagestreamlookup.go#L171

So we were incorrectly scoring an inexact match (perl:5.20) as an exact match, but because we were in a path where we assumed everything was an inexact match, we didn't provide metadata with it.

**Comment by bparees** [comment]
@jim-minter notice even the code comments imply we think these are all partial matches:
https://github.com/openshift/origin/blob/master/pkg/generate/app/imagestreamlookup.go#L147

**Comment by bparees** [comment]
> @bparees finally, I wonder whether this is worthy of a test case.
 
borderline (and i'm trying to get this in since it's a blocker for 3.5) but if i can come up w/ something simple and quick today i'll add it.

**Comment by bparees** [comment]
@jim-minter added a test case and also modified the code to simplify the scoring logic as we discussed.

**Comment by bparees** [comment]
[merge]

**Pull request merged**